### PR TITLE
Invoke() api

### PR DESCRIPTION
**Pull request by koba04**
This PR is to pick up from #945
If @milesj is still working on #945, feel free to close this!

### CODE REVIEW

**Comment by milesj** [comment]
I haven't had time to work on my PR, so thanks for this!

File: docs/api/ShallowWrapper/invoke.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+          Load more
+        </button>
+      </div>
+    );
+  }
+}
+
+const wrapper = shallow(<Foo />);
+
+wrapper.find('a').invoke('click').then(() => {
```
Comment (woutervanvliet) [inline docs/api/ShallowWrapper/invoke.md]: In the unit tests, invoke is used as below

```
wrapper.find('button').invoke('onClick')
```

Do they both work?

File: packages/enzyme-test-suite/test/ShallowWrapper-spec.jsx [hunk 2]
```diff
@@ -2440,6 +2440,70 @@ describe('shallow', () => {
...
+          );
+        }
+      }
+      const wrapper = shallow(<Foo />);
+      wrapper.find('button').invoke('onClick');
+      expect(wrapper.state('count')).to.equal(1);
+    });
+    it('can return the handlers return value', () => {
+      const spy = sinon.stub().returns(123);
+      class Foo extends React.Component {
```
Comment (woutervanvliet) [inline packages/enzyme-test-suite/test/ShallowWrapper-spec.jsx]: You can write shorter test by using a pure function component instead.

File: packages/enzyme-test-suite/test/ShallowWrapper-spec.jsx [hunk 3]
```diff
@@ -2440,6 +2440,70 @@ describe('shallow', () => {
...
+        render() {
+          return (
+            <div>
+              <a onClick={spy}>foo</a>
```
Comment (woutervanvliet) [inline packages/enzyme-test-suite/test/ShallowWrapper-spec.jsx]: Maybe change this to a button. There's a real risk that somebody will look at this code and use it as an example - and then are stuck with bad HTML semantics and potentially an app that doesn't work (don't think <a> tags are clickable without href)

File: packages/enzyme/src/ReactWrapper.js [hunk 4]
```diff
@@ -652,6 +652,26 @@ class ReactWrapper {
...
+   *
+   * @param {String} propName
+   * @param {Array} args
+   * @returns {Any}
+   */
+  invoke(propName, ...args) {
+    return this.single('invoke', () => {
+      const handler = this.prop(propName);
+      if (typeof handler !== 'function') {
+        throw new TypeError('ReactWrapper::invoke() expects a function name as its first argument');
```
Comment (woutervanvliet) [inline packages/enzyme/src/ReactWrapper.js]: Doesn't it expect a prop name as first argument?

**Comment by koba04** [comment]
@woutervanvliet Thank you for your reviews! I've fixed them.

File: docs/api/ShallowWrapper/invoke.md [hunk 5]
```diff
@@ -0,0 +1,41 @@
+# `.invoke(invokeName[, ...args]) => Any`
+
+Invokes a function prop.
+
+#### Arguments
+
+1. `propName` (`String`): The function prop to be invoked
+2. `...args` (`Any` [optional]): Arguments that will be passed to the prop function
```
Comment (ljharb) [inline docs/api/ShallowWrapper/invoke.md]: Using a spread here prevents ever being able to pass options.

The pattern we went with in `renderProp` was to return a new function, which could then be invoked with a variable number of arguments. What do you think about doing that here?

**Comment by ljharb** [comment]
@koba04 ping, still interested in this?

**Comment by koba04** [comment]
@ljharb OK, I'll work on it again!

**Comment by koba04** [comment]
I've recreated the branch from `master`.
Next, I'll adress https://github.com/airbnb/enzyme/pull/1856#discussion_r232826089

**Comment by koba04** [comment]
@ljharb I’ve addressed you comment and rebased this!

File: docs/api/ReactWrapper/invoke.md [hunk 6]
```diff
@@ -0,0 +1,41 @@
+# `.invoke(propName)(...args]) => Any`
```
Comment (ljharb) [inline docs/api/ReactWrapper/invoke.md]:
```suggestion
# `.invoke(propName) => (...args) => Any`
```

File: docs/api/ShallowWrapper/invoke.md [hunk 7]
```diff
@@ -0,0 +1,41 @@
+# `.invoke(invokePropName)(...args]) => Any`
```
Comment (ljharb) [inline docs/api/ShallowWrapper/invoke.md]:
```suggestion
# `.invoke(invokePropName) => (...args) => Any`
```

File: docs/api/mount.md [hunk 8]
```diff
@@ -159,6 +159,9 @@ Returns the props of the root component.
 #### [`.prop(key) => Any`](ReactWrapper/prop.md)
 Returns the named prop of the root component.
 
+#### [`.invoke(propName)(...args) => Any`](ReactWrapper/invoke.md)
```
Comment (ljharb) [inline docs/api/mount.md]:
```suggestion
#### [`.invoke(propName) => (...args) => Any`](ReactWrapper/invoke.md)
```

File: docs/api/shallow.md [hunk 9]
```diff
@@ -175,6 +175,9 @@ Returns the named prop of the current node.
 #### [`.key() => String`](ShallowWrapper/key.md)
 Returns the key of the current node.
 
+#### [`.invoke(propName)(...args) => Any`](ShallowWrapper/invoke.md)
```
Comment (ljharb) [inline docs/api/shallow.md]:
```suggestion
#### [`.invoke(propName) => (...args) => Any`](ShallowWrapper/invoke.md)
```

File: packages/enzyme-test-suite/test/ReactWrapper-spec.jsx [hunk 10]
```diff
@@ -5408,6 +5408,70 @@ describeWithDOM('mount', () => {
...
+              </button>
+            </div>
+          );
+        }
+      }
+      const wrapper = mount(<Foo />);
+      wrapper.find('button').invoke('onClick')();
+      expect(wrapper.state('count')).to.equal(1);
+    });
+    it('can return the handlers return value', () => {
```
Comment (ljharb) [inline packages/enzyme-test-suite/test/ReactWrapper-spec.jsx]:
```suggestion

    it('can return the handlers’ return value', () => {
```

File: packages/enzyme-test-suite/test/ReactWrapper-spec.jsx [hunk 11]
```diff
@@ -5408,6 +5408,70 @@ describeWithDOM('mount', () => {
...
+            </div>
+          );
+        }
+      }
+      const wrapper = mount(<Foo />);
+      const value = wrapper.find('a').invoke('onClick')();
+      expect(value).to.equal(123);
+      expect(spy.called).to.equal(true);
+    });
+    it('can pass in arguments', () => {
```
Comment (ljharb) [inline packages/enzyme-test-suite/test/ReactWrapper-spec.jsx]:
```suggestion

    it('can pass in arguments', () => {
```

File: packages/enzyme-test-suite/test/ShallowWrapper-spec.jsx [hunk 12]
```diff
@@ -5489,6 +5489,70 @@ describe('shallow', () => {
...
+              </button>
+            </div>
+          );
+        }
+      }
+      const wrapper = shallow(<Foo />);
+      wrapper.find('button').invoke('onClick')();
+      expect(wrapper.state('count')).to.equal(1);
+    });
+    it('can return the handlers return value', () => {
```
Comment (ljharb) [inline packages/enzyme-test-suite/test/ShallowWrapper-spec.jsx]:
```suggestion

    it('can return the handlers’ return value', () => {
```

File: packages/enzyme-test-suite/test/ShallowWrapper-spec.jsx [hunk 13]
```diff
@@ -5489,6 +5489,70 @@ describe('shallow', () => {
...
+            </div>
+          );
+        }
+      }
+      const wrapper = shallow(<Foo />);
+      const value = wrapper.find('button').invoke('onClick')();
+      expect(value).to.equal(123);
+      expect(spy.called).to.equal(true);
+    });
+    it('can pass in arguments', () => {
```
Comment (ljharb) [inline packages/enzyme-test-suite/test/ShallowWrapper-spec.jsx]:
```suggestion

    it('can pass in arguments', () => {
```

File: packages/enzyme/src/ReactWrapper.js [hunk 14]
```diff
@@ -802,6 +802,27 @@ class ReactWrapper {
...
+   * Will invoke an function prop and return its value.
+   *
+   * @param {String} propName
+   * @returns {Any}
+   */
+  invoke(propName) {
+    return this.single('invoke', () => {
+      const handler = this.prop(propName);
+      if (typeof handler !== 'function') {
+        throw new TypeError('ReactWrapper::invoke() expects a function prop name as its first argument');
```
Comment (ljharb) [inline packages/enzyme/src/ReactWrapper.js]:
```suggestion
        throw new TypeError('ReactWrapper::invoke() requires the name of a prop whose value is a function');
```

File: packages/enzyme/src/ReactWrapper.js [hunk 15]
```diff
@@ -802,6 +802,27 @@ class ReactWrapper {
...
+      }
+      return (...args) => {
+        const response = handler.apply(this, args);
```
Comment (ljharb) [inline packages/enzyme/src/ReactWrapper.js]: why should `this` be passed? if anything, it'd get `this.props`, but i think it'd be better to do:
```suggestion
        const response = handler(...args);
```

File: packages/enzyme/src/ReactWrapper.js [hunk 16]
```diff
@@ -802,6 +802,27 @@ class ReactWrapper {
...
+        this[ROOT].update();
```
Comment (ljharb) [inline packages/enzyme/src/ReactWrapper.js]: why is this needed?

File: packages/enzyme/src/ShallowWrapper.js [hunk 17]
```diff
@@ -1132,6 +1132,27 @@ class ShallowWrapper {
...
+   * Will invoke an function prop and return its value.
+   *
+   * @param {String} propName
+   * @returns {Any}
+   */
+  invoke(propName) {
+    return this.single('invoke', () => {
+      const handler = this.prop(propName);
+      if (typeof handler !== 'function') {
+        throw new TypeError('ShallowWrapper::invoke() expects a function prop name as its first argument');
```
Comment (ljharb) [inline packages/enzyme/src/ShallowWrapper.js]:
```suggestion
        throw new TypeError('ShallowWrapper::invoke() requires the name of a prop whose value is a function');
```

File: packages/enzyme/src/ShallowWrapper.js [hunk 18]
```diff
@@ -1132,6 +1132,27 @@ class ShallowWrapper {
...
+      }
+      return (...args) => {
+        const response = handler.apply(this, args);
```
Comment (ljharb) [inline packages/enzyme/src/ShallowWrapper.js]:
```suggestion
        const response = handler(...args);
```

File: packages/enzyme/src/ShallowWrapper.js [hunk 19]
```diff
@@ -1132,6 +1132,27 @@ class ShallowWrapper {
...
+        this[ROOT].update();
```
Comment (ljharb) [inline packages/enzyme/src/ShallowWrapper.js]: ditto

File: docs/api/ReactWrapper/invoke.md (hunk 6 above)
Comment (koba04) [reply docs/api/ReactWrapper/invoke.md]: @ljharb `invoke` is the same interface with `renderProp`, which is `.renderProp(key)() => ShallowWrapper`. Why is this change needed?
(`]` is a typo so I'll remove this though)

https://airbnb.io/enzyme/docs/api/shallow.html#renderpropkey--shallowwrapper

Comment (ljharb) [reply docs/api/ReactWrapper/invoke.md]: true, fair point. match renderProp for now; if i want to change them i'll change both at once later.

File: packages/enzyme/src/ReactWrapper.js (hunk 16 above)
Comment (koba04) [reply packages/enzyme/src/ReactWrapper.js]: @ljharb 
The purpose is to reflect the updates into the ShallowWrapper instance if the `handler` calls `setState`.

It's the same as what `simulate()` is doing.

https://github.com/airbnb/enzyme/blob/master/packages/enzyme/src/ShallowWrapper.js#L999

But I've covered the most case by this PR so this might be unnecessary.

https://github.com/airbnb/enzyme/pull/1499

**Comment by ljharb** [comment]
I'm not sure what's going on with coverage there; I'm going to merge anyways.

**Comment by koba04** [comment]
Thanks!
